pinc/ProofProject.inc

```diff
@@ -0,0 +1,365 @@
+            // Now they are 'unsaving' it, so decrement their page-count.
+            // They'll get it back if/when they save-as-done again.
+            page_tallies_add($this->round->id, $pguser, -1);
+        }
+        // otherwise it is in temp or out state: do nothing
+        return $this->render_page_data();
+    }
+
+    public function render_page_text(string $page_option): array
+    {
+        // page should be in out or temp state only
+        $this->validate_page(false, "getText");
+        return $this->get_page_text_array($page_option);
+    }
+
+    private function validate_page(bool $can_be_in_saved_state, string $op_name): void
+    {
+        $allowed_states = [$this->round->page_out_state, $this->round->page_temp_state];
+        if ($can_be_in_saved_state) {
+            $allowed_states[] = $this->round->page_save_state;
+        }
+        if (!in_array($this->page_state, $allowed_states)) {
+            $err = sprintf(_("The page state %1s is not allowed for this operation (%2s)."), $this->page_state, $op_name);
+            throw new ProjectPageStateException($err);
+        }
+    }
+}
+
+class TextChecker
+{
+    public function __construct(object $project)
+    {
+        $this->project = $project;
+        $this->pattern_string = build_character_regex_filter($project->get_valid_codepoints());
+    }
+
+    public const GOOD_TEXT = 0;
+    public const BAD_TEXT = 1;
+
+    private $text_array = [];
+    private $good_text_block = "";
+
+    public function analyse(): array
+    {
+        $this->receive_text();
+        $all_good = true;
+        foreach (split_graphemes($this->page_text) as $grapheme) {
+            if (1 === preg_match("/$this->pattern_string/u", $grapheme)) {
+                $this->good_text_block .= $grapheme;
+            } else {
+                $all_good = false;
+                $this->append_good();
+                $this->text_array[] = [$grapheme, self::BAD_TEXT];
+            }
+        }
+        // append any remaining good text
+        $this->append_good();
+        return [
+            "mark_array" => $this->text_array,
+            "good" => $all_good,
+        ];
+    }
+
+    private function append_good(): void
+    {
+        if ($this->good_text_block != "") {
+            $this->text_array[] = [$this->good_text_block, self::GOOD_TEXT];
+            $this->good_text_block = "";
+        }
+    }
+
+    public function get_valid_text(): string
+    {
+        $this->receive_text();
+        foreach (split_graphemes($this->page_text) as $grapheme) {
+            if (1 != preg_match("/$this->pattern_string/u", $grapheme)) {
+                throw new ProjectBadTextException(_("The text contains characters which are not allowed for this project"));
+            }
+        }
+        return _normalize_page_text($this->page_text, $this->project->projectid);
+    }
+
+    private function receive_text(): void
+    {
+        $request_data = api_get_request_body();
```

Comment:
This needs to be in the api and passed into the class. This is effectively user input which should always be handled by the top-most entity to ensure components, like this class, are reusable.

Reply:
done this